libiocage/lib/DevfsRules.py

```diff
@@ -0,0 +1,381 @@
+        """
+        return rule in self
+
+    def append(self, rule):
+        if rule not in self:
+            list.append(self, rule)
+
+    def clone(self, source_ruleset):
+        """
+        Clones the rules from another ruleset
+
+        Args:
+
+            source_ruleset (libiocage.lib.DevfsRules.DevfsRuleset):
+                Ruleset to copy all rules from
+        """
+        for rule in source_ruleset:
+            self.append(rule)
+
+    def _parse_line(self, line):
+
+        # marks beginning of a new ruleset
+        ruleset_match = re.search(DevfsRuleset.PATTERN, line)
+        if ruleset_match is not None:
+            name = str(ruleset_match.group("name"))
+            number = int(ruleset_match.group("number"))
+            comment = ruleset_match.group("comment")
+            return name, number, comment
+
+        raise SyntaxError("DevfsRuleset line parsing failed")
+
+    def __str__(self):
+        ruleset_line = f"[{self.name}={self.number}]"
+        if self.comment is not None:
+            ruleset_line += f" # {self.comment}"
+        output = [ruleset_line] + [str(x) for x in self]
+        return "\n".join(output) + "\n"
+
+
+class DevfsRules(list):
+    """
+    Abstraction for the hosts /etc/devfs.rules
+
+    Read and edit devfs rules in a programmatic way.
+    Restarts devfs service after applying changes.
+    """
+
+    def __init__(self, rules_file="/etc/devfs.rules", logger=None):
+        """
+        Initializes a DevfsRules manager for devfs.rules files
+
+        Args:
+
+            rules_file (string): (default=/etc/devfs.rules)
+                Path of the devfs.rules file
+
+            logger (libiocage.Logger): (optional)
+                Instance of the logger that is passed to occuring errors
+
+        """
+
+        self.logger=logger
+
+        # index rulesets to find duplicated and provide easy access
+        self._ruleset_number_index = {}
+        self._ruleset_name_index = {}
+
+        # all line indexes that came from default rules (here called system-rules)
+        self._system_rule_lines = []
+
+
+        list.__init__(self)
+
+        # will automatically read from file - needs to be the last item
+        self.rules_file = rules_file
+
+    def append(self, ruleset, is_system_rule=False):
+        """
+        Add a DevfsRuleset to the list
+
+        The rulesets added become indexed, so that lookups and duplication
+        checks are easy and fast
+
+        Args:
+
+            ruleset (libiocage.lib.DevfsRules.DevfsRuleset|string):
+                The ruleset that gets added if it is not already in the list
+        """
+
+        next_line_index = len(self)
+
+        if ruleset is None or isinstance(ruleset, str):
+            list.append(self, ruleset)
+            if is_system_rule is True:
+                self._system_rule_lines.append(next_line_index)
+            return ruleset
+
+        if ruleset.name in self._ruleset_name_index.keys():
```

Comment:
I think this and the last if below can be combined into an or

Reply:
Could be, but we need to still differentiate the message, so I'd think it would not look much cleaner after moving around. If you see more potential, please poke me again ;)